src/Core/SymbolUtility.cs

```diff
@@ -623,53 +624,66 @@ public static ulong GetEnumValueAsUInt64(object value, INamedTypeSymbol enumType
         }
     }
 
-    public static bool IsAwaitable(ITypeSymbol typeSymbol, bool shouldCheckWindowsRuntimeTypes = false)
+    public static bool IsWellKnownTaskType(this ITypeSymbol typeSymbol)
     {
-        INamedTypeSymbol? namedTypeSymbol = GetPossiblyAwaitableType(typeSymbol);
+        return typeSymbol.ContainingNamespace.HasMetadataName(in MetadataNames.System_Threading_Tasks)
+            && (typeSymbol.MetadataName is "Task" or "Task`1" or "ValueTask" or "ValueTask`1");
+    }
+
+    /// <summary>
+    /// Determines if the symbol is an awaitable type (i.e. the <see langword="await"/> keyword can be used on it) or a method that returns one.<br/>
+    /// A type is awaitable if it has an instance or extension <c>GetAwaiter</c> method that returns a correctly-shaped awaiter type.
+    /// </summary>
+    /// <remarks>
+    /// For more information, see the <see href="https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/expressions#12982-awaitable-expressions">C# language specification</see>.
+    /// </remarks>
+    /// <param name="semanticModel">Used to determine whether a candidate <c>GetAwaiter</c> method is accessible at the given <paramref name="position"/>.</param>
+    /// <param name="position">The position of the token at which the <paramref name="symbol"/> is used.</param>
+    /// <returns>A <see cref="bool"/> indicating whether the symbol is an awaitable type or a method that returns one.</returns>
+    public static bool IsAwaitable(this ISymbol? symbol, SemanticModel semanticModel, int position)
+    {
+        ITypeSymbol? typeSymbol = (symbol as ITypeSymbol) ?? (symbol as IMethodSymbol)?.ReturnType;
 
-        if (namedTypeSymbol is null)
+        if (typeSymbol is null or { SpecialType: SpecialType.System_Void })
             return false;
 
-        INamedTypeSymbol originalDefinition = namedTypeSymbol.OriginalDefinition;
-        TypeKind typeKind = originalDefinition.TypeKind;
+        if (typeSymbol.IsWellKnownTaskType())
```

Comment:
When this method is called from `IsAwaitableTaskType` we know that this check will never succeed.

Could you please add private overload of `IsAwaitable` which does not contain this check and will be called from `IsAwaitableTaskType`.